File: packages/shared/src/types/provider.ts
/**
 * Provider + MCP configuration contracts.
 *
 * Two distinct provider families on purpose:
 *  - AI providers         -> stream tokens + tool calls
 *  - Infrastructure providers -> expose Tools (github.*, aws.*, sentry.*)
 */

/**
 * Which protocol translates this provider. The complete vocabulary, in one place,
 * because it is the axis that reaches every layer of the app: the Zod schemas, the
 * Rust enum and the `provider_configs.kind` column, `runtime_provider_config()`, the
 * settings UI, and `buildProvider()` — the one place allowed to branch on it.
 *
 *  - `openai-compatible` — one adapter covering every endpoint that speaks the OpenAI
 *    shape (OpenAI, OpenRouter, Ollama, LM Studio, vLLM, in-house gateways).
 *  - `anthropic` / `gemini` — native protocols whose translation cannot be expressed
 *    as a compatible endpoint, so each has its own adapter (ADR 0011).
 *
 * Adding a kind here is not enough on its own: a kind that cannot be *saved* is a
 * kind the user cannot configure (ADR 0011 point 6).
 */
export const AI_PROVIDER_KINDS = ["openai-compatible", "anthropic", "gemini"] as const;

export type AiProviderKind = (typeof AI_PROVIDER_KINDS)[number];

/**
 * One AI provider row. `kind` and `wireApi` are **orthogonal axes** (ADR 0011 point 2):
 * `kind` says *who translates*, `wireApi` says *which dialect of that one translation*.
 * Adding a native kind must not leak `if (provider === "...")` into the loop.
 */
export interface AiProviderConfig {
  id: string;
  kind: AiProviderKind;
  label: string;
  baseUrl: string;
  model: string;
  /** Reference into the OS credential store. Never the key itself. */
  apiKeyCredentialRef?: string;
  enabled: boolean;
  /** Extra headers for corporate gateways. */
  customHeaders?: Record<string, string>;
  /**
   * Anthropic's dated `anthropic-version` protocol header.
   *
   * Only the native Anthropic adapter has this version axis. Other kinds must omit
   * it rather than store a value their protocol would silently ignore.
   */
  apiVersion?: string;
  maxInputTokens?: number;
  /**
   * Endpoint dialect (codex `responses` vs chat completions). **Only meaningful for
   * `openai-compatible`**: the two native kinds have no dialect axis, because the
   * protocol *is* the adapter's whole reason to exist.
   *
   * Both sides enforce that rather than ignoring the field: on a native kind the key
   * must be absent (`schemas/provider.ts` rejects it), and Rust omits it when it
   * serializes the row. So "present" always means "this is a chat-dialect choice",
   * and a `gemini` config carrying `wireApi: "responses"` is a rejected input instead
   * of a value whose meaning depends on whether the reader remembered to ignore it.
   */
  wireApi?: "chat" | "responses";
  /** Cached non-sensitive catalog entries used by the model selector. */
  models?: ProviderModelOption[];
  createdAt: string;
  updatedAt: string;
}

export interface ProviderModelOption {
  id: string;
  label: string;
  contextWindow?: number;
  supportsToolCalling: boolean;
  supportsStreaming: boolean;
}

/**
 * Per-run provider material resolved by Rust and injected with `agent.run.start`.
 * The API key rides only on this transient payload — never persisted, never
 * logged — while the durable config (baseUrl/model/label) lives in SQLite.
 */
export interface RuntimeProviderConfig {
  kind: AiProviderKind;
  /**
   * Full base URL. Rust falls back to the protocol's own public endpoint when the row
   * has none (or a blank one): `https://api.anthropic.com` for `anthropic`,
   * `https://generativelanguage.googleapis.com` for `gemini`. There is deliberately no
   * such fallback for `openai-compatible` — that kind covers endpoints we do not own,
   * so inventing a default would send the user's key somewhere they never chose.
   */
  baseUrl: string;
  model: string;
  /** Resolved at the point of use; absent for local endpoints (Ollama…). */
  apiKey?: string;
  customHeaders?: Record<string, string>;
  /**
   * Anthropic `anthropic-version` override. Only the native Anthropic kind may
   * carry it; see `AiProviderConfig.apiVersion`.
   */
  apiVersion?: string;
  timeoutMs?: number;
  /**
   * Endpoint dialect: chat completions (default) or the codex `responses` API.
   * Only `openai-compatible` may carry it — see `AiProviderConfig.wireApi`.
   */
  wireApi?: "chat" | "responses";
}

/** Settings form: label optional (defaults to baseUrl), apiKey goes to the keychain here. */
export interface ProviderSaveInput {
  providerId?: string;
  /**
   * Which protocol this row speaks. Required, not defaulted: the whole point of the
   * kind axis is that a stored provider says what it is, and a save that omits the kind
   * would have to be interpreted as one of the three (ADR 0011 point 6).
   */
  kind: AiProviderKind;
  label?: string;
  baseUrl: string;
  model: string;
  /** Present only when the user enters a new key; absent keeps the existing ref. */
  apiKey?: string;
  /** Non-secret gateway/protocol headers approved by the shared schema. */
  customHeaders?: Record<string, string>;
  /** Anthropic-only dated protocol version; see `AiProviderConfig.apiVersion`. */
  apiVersion?: string;
  /** Rejected for the two native kinds; see `AiProviderConfig.wireApi`. */
  wireApi?: "chat" | "responses";
  models?: ProviderModelOption[];
}

export interface ModelInfo {
  id: string;
  label: string;
  contextWindow?: number;
  supportsToolCalling: boolean;
  supportsStreaming: boolean;
}

export interface ProviderStatus {
  id: string;
  label: string;
  state: "connected" | "not_configured" | "error" | "disabled";
  models?: ModelInfo[];
  detail?: string;
}

/**
 * 删除一个 Provider 的结果。
 *
 * `credentialReclaimed`：那份 keychain 条目在删除后不再被任何 Provider 引用，于是被一并
 * 移除。引用是可以共享的（导入的数据里就有一份引用挂在四行上），所以「密钥还在不在」是
 * 这次操作唯一无法从别处看出来的后果。
 */
export interface ProviderDeleteResponse {
  deleted: boolean;
  credentialReclaimed: boolean;
}

/** — infrastructure providers contribute tools, they are not called directly. */
export interface InfrastructureProviderConfig {
  id: string;
  kind: "github" | "gitlab" | "aws" | "gcp" | "azure" | "cloudflare" | "vercel" | "sentry" | "datadog" | "kubernetes";
  label: string;
  credentialRef?: string;
  enabled: boolean;
  settings?: Record<string, unknown>;
}

/** — MCP servers are untrusted by default. */
export interface McpServerConfig {
  id: string;
  label: string;
  transport: "stdio" | "http";
  command?: string;
  args?: string[];
  url?: string;
  /** Ordered static HTTP authentication headers; names are public, references stay opaque. */
  httpAuthHeaders: McpHttpAuthHeaderConfig[];
  /** OAuth discovery/client metadata and an opaque token-bundle reference. */
  oauth?: McpOAuthConfig;
  enabled: boolean;
  /**
   * Tools this server is allowed to register. Empty = nothing is auto-trusted;
   * the user must opt in per tool after seeing its description.
   */
  allowedTools: string[];
  /** Every MCP tool starts at >= "medium" until reviewed. */
  trustLevel: "reviewed" | "unreviewed";
}

/** One stored static HTTP authentication header. */
export interface McpHttpAuthHeaderConfig {
  name: string;
  /** Opaque credential-store reference; the secret itself never reaches this contract. */
  credentialRef: string;
}

/**
 * How the desktop obtains the first token bundle (mirrors `McpOAuthFlow`).
 *
 * `authorization_code` is the browser redirect with a loopback callback; `device_code`
 * is RFC 8628, for servers that cannot reach a local callback at all.
 */
export type McpOAuthFlow = "authorization_code" | "device_code";

/**
 * Client authentication at the token endpoint (mirrors `McpOAuthClientAuth`).
 *
 * `none` is a public client. A client id provisioned by dynamic registration is always
 * `none`: a secret the server volunteers during registration is not a reason to start
 * sending one.
 */
export type McpOAuthClientAuth = "none" | "client_secret_post" | "client_secret_basic";

/** Stored OAuth configuration. Access and refresh tokens never enter this shape. */
export interface McpOAuthConfig {
  issuer: string;
  clientId: string;
  /**
   * Always present: Rust defaults rows written before the field existed to
   * `authorization_code`, and a flow change is part of the stored identity, so
   * switching it discards the stored token rather than reusing it.
   */
  flow: McpOAuthFlow;
  /** Always present: rows written before this setting existed are public clients. */
  clientAuth: McpOAuthClientAuth;
  /**
   * Opaque credential-store reference for the hand-entered client secret. The secret
   * itself never reaches this contract, a settings response, an error or a log.
   */
  clientSecretRef?: string;
  /**
   * Whether access tokens must be sender-constrained (RFC 9449 DPoP, ADR 0018).
   *
   * Always present: rows written before the setting existed are plain bearer clients, and
   * turning it on or off is part of the stored identity — it discards the stored token
   * exactly like changing the issuer does.
   */
  dpop: boolean;
  /**
   * Opaque credential-store reference for the DPoP private key. The key is generated on
   * the host and never leaves the credential store, so losing it costs a re-authorization
   * instead of silently downgrading the connection to bearer.
   */
  dpopKeyRef?: string;
  scopes: string[];
  tokenEndpoint?: string;
  credentialRef?: string;
}

File: crates/database/src/models/provider.rs
//! provider configs (AI + infrastructure) and MCP servers.

use serde::{Deserialize, Serialize};

/// Non-sensitive model metadata cached from a provider catalog. The API key is
/// deliberately absent so this value is safe to persist in SQLite and return
/// to the desktop UI.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderModelOption {
    pub id: String,
    pub label: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_window: Option<u64>,
    pub supports_tool_calling: bool,
    pub supports_streaming: bool,
}

// `kind` 是 `provider_configs` 的一列，所以它属于这张表：wire 拼写与列里的拼写是同一份。
enum_as_str!(AiProviderKind, OpenaiCompatible => "openai-compatible", Anthropic => "anthropic", Gemini => "gemini");

/// Provider 身份（ADR 0011）：`kind` 决定**由谁翻译**，与 `AiProviderConfig::wire_api`
/// 那条「怎么翻译」的轴正交。三个取值必须一路走到数据库列、IPC 与设置界面 —— 一个存不进去
/// 或读不回来的 kind，等于「实现了一个用户配不出来的 Provider」。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum AiProviderKind {
    /// 写出去的一律是前端契约里的 `openai-compatible`；`openaicompatible` 是旧的
    /// `rename_all = "lowercase"` 曾经落盘的拼写，`kind` 是 TEXT 列且没有迁移改写它，
    /// 所以只放宽读取，避免升级后旧行直接反序列化失败。
    #[serde(alias = "openaicompatible")]
    OpenaiCompatible,
    /// Anthropic Messages API（原生协议，`anthropic.ts`）。
    Anthropic,
    /// Google Gemini `generateContent`（原生协议，`gemini.ts`）。
    Gemini,
}

impl AiProviderKind {
    /// `provider_configs.kind` 列的读入。比 serde 多一件事：旧拼写 `openaicompatible`
    /// 也在这里被接受（同一条「只放宽读取」的规则，两条路径不能分叉）。
    #[must_use]
    pub fn from_db_column(raw: &str) -> Option<Self> {
        Self::from_db(raw).or_else(|| (raw == "openaicompatible").then_some(Self::OpenaiCompatible))
    }

    /// 这个 kind 是否有 `wireApi` 方言轴。**只有 OpenAI-compatible 有**：原生协议的
    /// 翻译本身就是它唯一的形式（ADR 0011 第 2 点），给它配一个方言是无意义的。
    #[must_use]
    pub fn has_wire_api(self) -> bool {
        matches!(self, Self::OpenaiCompatible)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiProviderConfig {
    pub id: String,
    pub kind: AiProviderKind,
    pub label: String,
    pub base_url: String,
    pub model: String,
    /// "chat" | "responses" — 由用户按服务商端点选择。**只对 `OpenaiCompatible` 有意义**：
    /// 原生协议没有方言可选，所以那两种 kind 这里是 `None`，序列化时整个省略（界面因此
    /// 不会给它们显示一个「Wire API」），共享 schema 也把「原生 kind 带 wireApi」判为非法输入。
    ///
    /// 列是 `NOT NULL DEFAULT 'chat'`（迁移 2），写不进 NULL，所以 `None` 落的是空串；
    /// 读回来时由 `kind` 决定这一列是否被读（见 `repositories/providers.rs`）。
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub wire_api: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub api_key_credential_ref: Option<String>,
    pub enabled: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom_headers: Option<serde_json::Map<String, serde_json::Value>>,
    /// Anthropic 的日期版本头。只有 native Anthropic 协议使用它，其他 kind 必须是
    /// `None`；共享 schema 与保存命令都会拒绝越界组合。
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub api_version: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_input_tokens: Option<u32>,
    /// Cached model catalog. Stored under the existing `settings` column to
    /// keep the migration backwards compatible with existing databases.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub models: Option<Vec<ProviderModelOption>>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InfrastructureProviderConfig {
    pub id: String,
    pub kind: String,
    pub label: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub credential_ref: Option<String>,
    pub enabled: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub settings: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct McpServerConfig {
    pub id: String,
    pub label: String,
    /// "stdio" | "http"
    pub transport: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub command: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub args: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    #[serde(default)]
    pub http_auth_headers: Vec<McpHttpAuthHeaderConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub oauth: Option<McpOAuthConfig>,
    pub enabled: bool,
    pub allowed_tools: Vec<String>,
    /// "reviewed" | "unreviewed"
    pub trust_level: String,
}

/// One ordered static HTTP authentication header. The name is public; the
/// credential reference resolves through the OS credential store.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct McpHttpAuthHeaderConfig {
    pub name: String,
    pub credential_ref: String,
}

/// Non-secret OAuth configuration plus an opaque reference to the token bundle.
/// The token endpoint is cached after discovery so request-time refresh does not
/// need to fetch metadata again.
///
/// `flow` is part of the stored identity: switching between the browser redirect and
/// the device-code flow changes which requests the authorization server saw, so it
/// invalidates a stored token exactly like changing the client id does
/// (`commands/mcp.rs::resolve_oauth_config`).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct McpOAuthConfig {
    pub issuer: String,
    pub client_id: String,
    #[serde(default)]
    pub flow: McpOAuthFlow,
    /// How this client authenticates at the token endpoint.
    ///
    /// A dynamically registered client is always `none`: the server may hand back a
    /// secret in its registration response, and using it would silently turn a public
    /// client into one that only works while that extra value survives.
    #[serde(default)]
    pub client_auth: McpOAuthClientAuth,
    /// Credential-store reference for the hand-entered client secret. The secret itself
    /// never enters SQLite.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub client_secret_ref: Option<String>,
    /// Whether access tokens must be sender-constrained (RFC 9449 DPoP).
    ///
    /// Off by default: only a server that actually validates proofs benefits, and a client
    /// that sends them to one that does not has only added a header. When it is on, the
    /// token response must say `token_type: DPoP` — see ADR 0018.
    #[serde(default)]
    pub dpop: bool,
    /// Credential-store reference for the DPoP private key. The key itself never enters
    /// SQLite, and losing it costs a re-authorization rather than a silent downgrade.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dpop_key_ref: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub scopes: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token_endpoint: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub credential_ref: Option<String>,
}

/// How the desktop obtains the first token bundle.
///
/// Both flows end in the same refreshable bundle and the same credential-store entry;
/// they differ only in how the user proves they are present. Kept as a closed enum so
/// an unknown value in a stored row or an IPC payload is refused instead of silently
/// falling back to one of them.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum McpOAuthFlow {
    /// Redirect to the authorization endpoint and listen on a loopback callback.
    #[default]
    AuthorizationCode,
    /// RFC 8628: show a `user_code`, then poll the token endpoint.
    DeviceCode,
}

/// Client authentication at the token endpoint (RFC 6749 §2.3).
///
/// `none` is a public client: it sends only its `client_id`. The two secret methods are
/// the ones a server can actually distinguish, and the choice has to be explicit because
/// sending a secret the server did not ask for is a credential leak with no upside.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum McpOAuthClientAuth {
    /// Public client: `client_id` in the form, no secret anywhere.
    #[default]
    None,
    /// `client_id` and `client_secret` as form parameters.
    ClientSecretPost,
    /// `client_id:client_secret` in the `Authorization` header, and nothing in the form.
    ClientSecretBasic,
}

impl McpOAuthClientAuth {
    /// Does this method require a stored secret?
    pub fn needs_secret(self) -> bool {
        !matches!(self, Self::None)
    }
}
